### PR TITLE
Check template expression length #1006

### LINKED ISSUE #1006: Check template expressions are less or equal to maximum limit
# Rule request

## Suggested rule change

Previously PSRule for Azure enforced a maximum limit for template expansion in line with ARM documented limits, 24576 characters. This is the recommended best practise however it is no longer feasible to implement this as strict as previous due Bicep ARM generation.

By implementing this as a rule, customers can ignore the rule instead of it being a hard limit for PSRule for Azure. Also, the rule can selectively not apply to Bicep generated templates.

See related issue #1005.

## Applies to the following

The rule applies to the following:

- **Templates files**

## Additional context

[Reference](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/best-practices#template-limits)

### PR DESCRIPTION
**Pull request by BernieWhite**
## PR Summary

- Check template expressions do not exceed a maximum length.

Fixes #1006

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [x] Unit tests created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: docs/en/rules/Azure.Template.ExpressionLength.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+# Template expressions should not exceed a maximum length
+
+## SYNOPSIS
+
+Use template expressions should not exceed the maximum length.
+
+## DESCRIPTION
+
+Extemely long expressions may be difficult to read and debug.
```
Comment (ArmaanMcleod) [inline docs/en/rules/Azure.Template.ExpressionLength.md:18]:
```suggestion
Extremely long expressions may be difficult to read and debug.
```

File: docs/en/rules/Azure.Template.ExpressionLength.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+category: Repeatable infrastructure
+resource: All resources
+online version: https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.Template.ExpressionLength/
+---
+
+# Template expressions should not exceed a maximum length
+
+## SYNOPSIS
+
+Use template expressions should not exceed the maximum length.
```
Comment (ArmaanMcleod) [inline docs/en/rules/Azure.Template.ExpressionLength.md:14]:
```suggestion
Template expressions should not exceed the maximum length.
```

File: src/PSRule.Rules.Azure/rules/Azure.Template.Rule.ps1 [hunk 3]
```diff
@@ -236,6 +236,11 @@ Rule 'Azure.Template.ParameterStrongType' -Type 'Microsoft.Resources/deployments
     $Assert.Create($PSRule.Issue.Get('PSRule.Rules.Azure.Template.ParameterStrongType'));
 }
 
+# Synopsis: Use template expressions should not exceed the maximum length.
```
Comment (ArmaanMcleod) [inline src/PSRule.Rules.Azure/rules/Azure.Template.Rule.ps1:239]:
```suggestion
# Synopsis: Template expressions should not exceed the maximum length.
```

File: src/PSRule.Rules.Azure/Data/Template/TemplateVisitor.cs [hunk 4]
```diff
@@ -150,6 +155,9 @@ internal TemplateContext(PipelineContext context, SubscriptionOption subscriptio
 
             public ExpressionFnOuter BuildExpression(string s)
             {
+                if (s != null && s.Length > 24576 && !IsGenerated())
+                    AddValidationIssue(ISSUE_PARAMETER_EXPRESSIONLENGTH, s, ReasonStrings.ExpressionLength, s, 24576);
```
Comment (ArmaanMcleod) [inline src/PSRule.Rules.Azure/Data/Template/TemplateVisitor.cs:159]: Maybe we could put `24575` in a constant?

File: src/PSRule.Rules.Azure/rules/Azure.Template.Rule.ps1 (hunk 3 above)
Comment (ArmaanMcleod) [reply src/PSRule.Rules.Azure/rules/Azure.Template.Rule.ps1:239]: Just adding same suggestion as above.

**Review by ArmaanMcleod — APPROVED** [review]
@BernieWhite looks great. I just added some suggestions that might increase clarity 🙂

**Comment by BernieWhite** [comment]
@ArmaanMcleod Thanks, great suggestions.

**Pull request merged**